### PR TITLE
【AutoParallel】Add c_embedding pass in PIR

### PR DESCRIPTION
**Pull request by Xing-lil**
### PR Category

Auto Parallel

### PR Types

New features

### Description

add c_embedding pass in pir
Pcard-70448

### CODE REVIEW

File: python/paddle/distributed/passes/auto_parallel_c_embedding.py [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+                )
+
+
+@register_pass("auto_parallel_c_embedding_pass")
+class AutoParallelCEmbeddingPass(PassBase):
+    def __init__(self):
+        super().__init__()
+
+    def _check_self(self):
+        return True
```
Comment (liym27) [inline python/paddle/distributed/passes/auto_parallel_c_embedding.py]: Returned in advance

File: python/paddle/distributed/passes/auto_parallel_c_embedding.py [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+
+    def _apply_single_impl(self, main_program, startup_program, context):
+        concrete_program = self.get_attr("concrete_program")
+        ops = main_program.global_block().ops
+        for i, op in enumerate(ops):
+            if op.name() == 'pd_op.embedding':
+                # update weight dims mapping
+                mp_axis = update_weight(op, concrete_program)
+
+                # update startup_program
```
Comment (liym27) [inline python/paddle/distributed/passes/auto_parallel_c_embedding.py]: Meaningless comments.   delete line 334, 347, 350, 353...

**Review by liym27 — COMMENTED** [review]
Add reasonable comments to pass

File: python/paddle/distributed/passes/auto_parallel_c_embedding.py (hunk 1 above)
Comment (Xing-lil) [reply python/paddle/distributed/passes/auto_parallel_c_embedding.py]: Thx, the code has been modified.

File: python/paddle/distributed/passes/auto_parallel_c_embedding.py (hunk 2 above)
Comment (Xing-lil) [reply python/paddle/distributed/passes/auto_parallel_c_embedding.py]: Thx, the code has been modified.

File: paddle/phi/infermeta/spmd_rules/c_embedding.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+SpmdInfo CEmbeddingGradInferSpmd(const DistMetaTensor& weight,
+                                 const DistMetaTensor& x,
+                                 const DistMetaTensor& out_grad,
+                                 int start_index);
+
+SpmdInfo CEmbeddingGradInferSpmdBase(const DistMetaTensor& weight,
+                                     const DistMetaTensor& x,
+                                     const DistMetaTensor& out_grad,
+                                     int64_t start_index);
```
Comment (liym27) [inline paddle/phi/infermeta/spmd_rules/c_embedding.h]: Remove redundant functions `CEmbeddingGradInferSpmdBase` and `CEmbeddingInferSpmdBase`

File: paddle/phi/infermeta/spmd_rules/c_embedding.cc [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+          << str_join(w_dst.dist_attr().dims_mapping()) << "]\n"
+          << "Input out_grad shape: ["
+          << str_join(phi::vectorize(out_grad.dims()))
+          << "], src_dims_mapping: ["
+          << str_join(out_grad.dist_attr().dims_mapping())
+          << "], dst_dims_mapping: ["
+          << str_join(out_grad_dst.dist_attr().dims_mapping()) << "]\n"
+          << "Output w_grad shape: [" << str_join(phi::vectorize(w_grad.dims()))
+          << "], dims_mapping: [" << str_join(w_grad.dist_attr().dims_mapping())
+          << "]\n\n";
```
Comment (liym27) [inline paddle/phi/infermeta/spmd_rules/c_embedding.cc]: Replace line222-240 with functions `LOG_SPMD_INPUT`  and `LOG_SPMD_OUTPUT`

File: paddle/phi/infermeta/spmd_rules/c_embedding.cc [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+  VLOG(4) << "CEmbeddingInferSpmdBase:\n"
+          << "Einsum notation: [" << x_axes << "," << weight_axes << "]. "
+          << "X shape: [" << str_join(x_shape) << "], src_dims_mapping: ["
+          << str_join(x_dims_mapping) << "], dst_dims_mapping: ["
+          << str_join(x_dist_attr_dst.dims_mapping()) << "]\n W shape: ["
+          << str_join(weight_shape) << "], src_dims_mapping: ["
+          << str_join(weight_dims_mapping) << "], dst_dims_mapping: ["
+          << str_join(weight_dist_attr_dst.dims_mapping())
+          << "]\n Out dims_mapping: [" << str_join(out_dims_mapping)
+          << "], partial_on_dims: [" << str_join(partial_on_dims) << "]\n\n";
```
Comment (liym27) [inline paddle/phi/infermeta/spmd_rules/c_embedding.cc]: Replace line121-130 with functions `LogInputDistAttr`  and `LogOutputDistAttr`

File: paddle/phi/infermeta/spmd_rules/c_embedding.cc [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+                        "The Tensor out_grad's rank [%d] and out_grad's "
+                        "dims_mapping size [%d] are not matched.",
+                        out_grad.dims(),
+                        out_grad.dist_attr().dims_mapping().size()));
+  // Propagate sharding info using composite operators.
+  // The whole mathematical expression of CEmbeddingGrad is:
+  // w_grad = einsum('...j, ...k->jk', onehot(x, j), out_grad)
+
+  // TODO(cxxly): Simplifies the code logic of sharding propagation using
+  // primitive operators.
```
Comment (liym27) [inline paddle/phi/infermeta/spmd_rules/c_embedding.cc:123]: why TODO**(cxxly)**

File: test/auto_parallel/pir/auto_parallel_c_embedding_pass.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+            learning_rate=0.1, parameters=dy_layer.parameters()
+        )
+        dy_losses = self.run_dynamic(dy_layer, dy_opt, data_loader)
+
+        np.testing.assert_allclose(
+            dy2static_losses_use_pass, dy2static_losses, atol=1e-7
+        )
+        np.testing.assert_allclose(
+            dy2static_losses_use_pass, dy_losses, atol=1e-7
+        )
```
Comment (liym27) [inline test/auto_parallel/pir/auto_parallel_c_embedding_pass.py]: 1. Require equal md5 values
2. Check the program contains `c_embedding` op when using pass

File: test/auto_parallel/pir/test_auto_parallel_c_embedding_pass.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        )
+        self._default_envs = {"dtype": "float32", "seed": "2024"}
+        self._changeable_envs = {"backend": ["gpu"]}
+
+    def test_mlp(self):
+        envs_list = test_base.gen_product_envs_list(
+            {"dtype": "float32", "seed": "2024"}, {"backend": ["gpu"]}
+        )
+        for envs in envs_list:
+            # self._log_dir.name = "./log"
```
Comment (liym27) [inline test/auto_parallel/pir/test_auto_parallel_c_embedding_pass.py]: Delete redundant code

File: test/auto_parallel/spmd_rules/test_c_embedding_rule.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+            process_mesh  # not set the dims mapping is ok.
+        )
+        self.table_dist_tensor_spec = DistTensorSpec(
+            table_shape, table_tensor_dist_attr
+        )
+
+        x_tensor_dist_attr = TensorDistAttr()
+        x_tensor_dist_attr.process_mesh = (
+            process_mesh  # not set the dims mapping is ok.
+        )
```
Comment (liym27) [inline test/auto_parallel/spmd_rules/test_c_embedding_rule.py]: L104-L106 ->  `x_tensor_dist_attr.process_mesh = process_mesh`

File: test/auto_parallel/spmd_rules/test_c_embedding_rule.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        # backward setup
+        process_mesh = auto.ProcessMesh(mesh=[[0, 1, 2, 3], [4, 5, 6, 7]])
+
+        table_shape = [512, 768]  # [V,H]
+        x_shape = [4, 1024]  # [B,S]
+
+        table_tensor_dist_attr = TensorDistAttr()
+        table_tensor_dist_attr.process_mesh = (
+            process_mesh  # not set the dims mapping is ok.
+        )
```
Comment (liym27) [inline test/auto_parallel/spmd_rules/test_c_embedding_rule.py]: `table_tensor_dist_attr.process_mesh = process_mesh`

Comment (Xing-lil) [reply test/auto_parallel/spmd_rules/test_c_embedding_rule.py]: Thx, the code has been modified according to the suggestions above.

**Review by liym27 — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]

**Pull request merged**